src/_secureaccesstokenauthorizer/_storage_server.py

```diff
@@ -0,0 +1,169 @@
+    :param list[(bytes, foolscap.IConstraint)] kwargs: The arguments to
+        prepend to the signature of ``schema``.
+
+    :return foolscap.remoteinterface.RemoteMethodSchema: The new schema
+        object.
+    """
+    new_kwargs = dict(schema.argConstraints)
+    new_kwargs.update(kwargs)
+    modified_schema = RemoteMethodSchema(**new_kwargs)
+    # Initialized from **new_kwargs, RemoteMethodSchema.argumentNames is in
+    # some arbitrary, probably-incorrect order.  Fix it.
+    modified_schema.argumentNames = (
+        list(argName for (argName, _) in kwargs) +
+        schema.argumentNames
+    )
+    return modified_schema
+
+
+
+class RITokenAuthorizedStorageServer(RemoteInterface):
+    __remote_name__ = (
+        "RITokenAuthorizedStorageServer.tahoe.privatestorage.io"
+    )
+
+    get_version = RIStorageServer["get_version"]
+
+    allocate_buckets = add_tokens(RIStorageServer["allocate_buckets"])
+
+    add_lease = add_tokens(RIStorageServer["add_lease"])
+
+    renew_lease = add_tokens(RIStorageServer["renew_lease"])
+
+    get_buckets = RIStorageServer["get_buckets"]
+
+    slot_readv = RIStorageServer["slot_readv"]
+
+    slot_testv_and_readv_and_writev = add_tokens(
+        RIStorageServer["slot_testv_and_readv_and_writev"],
+    )
+
+    advise_corrupt_share = RIStorageServer["advise_corrupt_share"]
+
+
+
+@implementer_only(RITokenAuthorizedStorageServer, IReferenceable, IRemotelyCallable)
+# It would be great to use `frozen=True` (value-based hashing) instead of
+# `cmp=False` (identity based hashing) but Referenceable wants to set some
+# attributes on self and it's hard to avoid that.
+@attr.s(cmp=False)
+class SecureAccessTokenAuthorizerStorageServer(Referenceable):
+    _original = attr.ib()
+
+    def _validate_tokens(self, tokens):
+        pass
+
+    def remote_allocate_buckets(self, tokens, *a, **kw):
+        self._validate_tokens(tokens)
+        return self._original.remote_allocate_buckets(*a, **kw)
+
+    def remote_get_buckets(self, storage_index):
+        return self._original.remote_get_buckets(storage_index)
+
+    def remote_add_lease(self, tokens, *a, **kw):
+        self._validate_tokens(tokens)
+        return self._original.remote_add_lease(*a, **kw)
+
+    def remote_renew_lease(self, tokens, *a, **kw):
+        self._validate_tokens(tokens)
+        return self._original.remote_renew_lease(*a, **kw)
+
+    def remote_advise_corrupt_share(self, *a, **kw):
+        return self._original.remote_advise_corrupt_share(*a, **kw)
+
+    def remote_slot_testv_and_readv_and_writev(self, tokens, *a, **kw):
+        self._validate_tokens(tokens)
+        return self._original.remote_slot_testv_and_readv_and_writev(*a, **kw)
+
+    def remote_slot_readv(self, *a, **kw):
+        return self._original.remote_slot_readv(*a, **kw)
+
+# I don't understand why this is required.
+# SecureAccessTokenAuthorizerStorageServer is-a Referenceable.  It seems like
+# the built in adapter should take care of this case.
```

Comment:
Is it because you said `implementer_only` instead of `implementer`...? (also just guessing)

Reply:
oh yea maybe because I didn't include `providedBy(Referenceable)` in the list of things it only implements...

Reply:
`providedBy(Referenceable())` or `implementedBy(Referenceable)`, of course.  Or maybe `directlyProvidedBy(Referenceable())` / `directlyImplementedBy(Referenceable)`.

Reply:
Apparently there is no `directlyImplementedBy` and adding `directlyProvidedBy(Referenceable())` to the `implements_only` list doesn't make a difference.  So I dunno.